### PR TITLE
Gapi sycl backend

### PR DESCRIPTION
**Pull request by Blair-Johnson**

### CODE REVIEW

File: modules/gapi/CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,17 @@ ocv_add_module(gapi
       python
 )
 
+if(WITH_SYCL)
+  set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_LIST_DIR}/cmake")
+  find_package(SYCL REQUIRED)
+  if(SYCL_FOUND)
+    set(HAVE_SYCL TRUE)
+    set(CMAKE_CXX_COMPILER ${SYCL_COMPILER})
+    set(CMAKE_CXX_STANDARD 17)
```
Comment (dmatveev) [inline modules/gapi/CMakeLists.txt:37]: Is this required to support SYCL? It may gate the whole integration as OpenCV is so far C++11. We can always check it with the OpenCV team, though

File: modules/gapi/cmake/FindSYCL.cmake [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  VERSION_VAR SYCL_LANGUAGE_VERSION
+  REASON_FAILURE_MESSAGE "${SYCL_REASON_FAILURE}")
+
+# Include in Cache
+set(SYCL_LANGUAGE_VERSION "${SYCL_LANGUAGE_VERSION}" CACHE STRING "SYCL Language version")
+set(SYCL_INCLUDE_DIR "${SYCL_INCLUDE_DIR}" CACHE FILEPATH "SYCL Include directory")
+set(SYCL_LIBRARY_DIR "${SYCL_LIBRARY_DIR}" CACHE FILEPATH "SYCL Library Directory")
+set(SYCL_FLAGS "${SYCL_FLAGS}" CACHE STRING "SYCL flags for the compiler")
+set(SYCL_COMPILER "${SYCL_COMPILER}" CACHE STRING "SYCL compiler path")
+set(SYCL_LIBRARY "${SYCL_LIBRARY}" CACHE STRING "SYCL library")
```
Comment (dmatveev) [inline modules/gapi/cmake/FindSYCL.cmake:252]: Where does this file comes from (re LICENSE)? Probably it worth adding it at the OpenCV level as it may be useful in other modules.
But it can always be done later, of course.

File: modules/gapi/include/opencv2/gapi/sycl/core.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <opencv2/core/cvdef.h> // GAPI_EXPORTS
+#include <opencv2/gapi/gkernel.hpp> // GKernelPackage
+
+namespace cv {
+namespace gapi {
+namespace core {
+namespace sycl {
+
+  GAPI_EXPORTS_W cv::gapi::GKernelPackage kernels();
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/core.hpp]: In OpenCV namespaces don't add extra indent (we learned about this rule much later so our code may be not ideal w.r.t this rule. :) )

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+#include <map>
+#include <unordered_map>
+
+#include <opencv2/core/mat.hpp>
+#include <opencv2/gapi/gcommon.hpp>
+#include <opencv2/gapi/gkernel.hpp>
+#include <opencv2/gapi/garg.hpp>
+
+// FIXME: These are not included with OpenCV by default
+// This is protected by HAVE_SYCL in gsyclbackend.hpp, is this included anywhere else?
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: public headers shouldn't refer to `HAVE_` compile flags as those are never set when a user app includes those.

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+#include <CL/sycl.hpp>
+#include <CL/sycl/backend/opencl.hpp>
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: Do we need these headers here or we can keep forward declarations only?

I believe user will include it himself if he/she plans to write SYCL code

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+// FIXME: These are not included with OpenCV by default
+// This is protected by HAVE_SYCL in gsyclbackend.hpp, is this included anywhere else?
+#include <CL/sycl.hpp>
+#include <CL/sycl/backend/opencl.hpp> 
+
+// FIXME: namespace scheme for backends?
+namespace cv {
+
+    namespace gimpl
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: please avoid namespace indents

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        sycl::queue& getQueue();
+
+        // Generic accessor API
+        template<typename T>
+        const T& inArg(int input) { return m_args.at(input).get<T>(); }
+
+        // Syntax sugar
+        const sycl::buffer<uint8_t, 2>& inMat(int input);
+        sycl::buffer<uint8_t, 2>& outMatR(int output);
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: probably forward template declarations work for this.. need to try it out

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        template<typename T> T& outOpaqueR(int output) // FIXME: the same issue
+        {
+            return outOpaqueRef(output).wref<T>();
+        }
+
+    protected:
+        // SYCL specific values
+        // TODO: Determine when these get assigned
+        sycl::queue m_queue;
+        sycl::context m_context;
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: Since the queue and context should be owned by backend (providing access kernels with access to those via this context), I believe these members should be references?

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+        void initSYCLContext();
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: is it correct? I believe context is associated with (sub)graph but not with individual kernel?

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    };
+
+    // FIXME: This is an ugly ad-hoc implementation. TODO: refactor
+    // TODO: Figure out the difference between the gtyped input wrappers and the cv typed ones
+    namespace detail
+    {
+        template<class T> struct sycl_get_in;
+        template<> struct sycl_get_in<cv::GMat>
+        {
+            static const sycl::buffer<uint8_t, 2> get(GSYCLContext& ctx, int idx)
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: btw what is `2` here? is it a number of dimensions? What if the mat is tensor (e.g. has 4-5 dims)?

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: Also, does it represent well multi-channel images (e.g. CV_8UC3)?

File: modules/gapi/include/opencv2/gapi/util/variant.hpp [hunk 11]
```diff
@@ -25,23 +25,27 @@ namespace util
         template<std::size_t I, typename Target, typename First, typename... Remaining>
         struct type_list_index_helper
         {
-            static const constexpr bool is_same = std::is_same<Target, First>::value;
-            static const constexpr std::size_t value =
+            // constexpr removed for C++17 compatibility (SYCL compiler)
+            static const bool is_same = std::is_same<Target, First>::value;
+            // constexpr removed for C++17 compatibility (SYCL compiler)
+            static const std::size_t value =
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/util/variant.hpp:31]: does SYCL compile this code or is it sensitive to these constructs?

Are these changes still required?

File: modules/gapi/src/backends/common/gbackend.hpp [hunk 12]
```diff
@@ -19,6 +19,10 @@
 
 #include "compiler/gmodel.hpp"
 
+#ifdef HAVE_SYCL
+#include <CL/sycl.hpp>
+#endif // HAVE_SYCL
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/gbackend.hpp]: This shouldn't be here in the common file, it is kept as platform-neutral

File: modules/gapi/src/backends/common/gbackend.hpp [hunk 13]
```diff
@@ -145,6 +149,9 @@ using Mag = magazine::Class< cv::Mat
 #if !defined(GAPI_STANDALONE)
                            , cv::UMat
 #endif
+#ifdef HAVE_SYCL //FIXME: Issues with sycl buffers being templated?
+                           , sycl::buffer<uint8_t, 2>
+#endif // HAVE_SYCL
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/gbackend.hpp]: Not sure if it should be there, normally we don't add backend-specific data structures to this generic container.

You can either define our own `magazine::Class<...>` in your backend (adding necessary type there), or use `RMat` as a data holder with associated `sycl::buffer` in its custom Adapter class.

File: modules/gapi/src/backends/sycl/gsyclbackend.cpp [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+          magazine::bindInArg(m_res, rc, m_gm.metadata(nh).get<ConstValue>().arg);
+        }
+        //This is where internal mat allocation happens
+        //need to figure out how to use buffers instead
+        if (desc.storage == Data::Storage::INTERNAL && desc.shape == GShape::GMAT)
+        {
+          const auto mat_desc = util::get<cv::GMatDesc>(desc.meta);
+          //auto& mat = m_res.slot<cv::Mat>()[desc.rc];
+          //createMat(mat_desc, mat);
+        }
```
Comment (dmatveev) [inline modules/gapi/src/backends/sycl/gsyclbackend.cpp]: Here you can create an RMat with sycl buffer inside and put it in the magazine under the given ID.

File: modules/gapi/src/backends/sycl/gsyclbackend.cpp [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+        CV_CheckTypeEQ(mag_mat.type(), CV_8UC1, "Only single channel UInt8 Mats " <<
+                                    "are currently supported by SYCL");
+        mag_buffer(mag_mat.data, range<2>(mag_mat.rows, mag_mat.cols));
+    };
+
+    for (auto& it : input_objs) {
+        const auto& rc = it.first;
+        magazine::bindInArg(m_res, rc, it.second);
+        // There is already a cv::Mat in the magazine after bindInArg call,
+        // extract buffer from it and put into magazine
```
Comment (dmatveev) [inline modules/gapi/src/backends/sycl/gsyclbackend.cpp]: the default `bindInArg` implementation is old (left here for backwards compatibility with the existing backends).
Please call to make `RMat` be stored in magazine instead of `cv::Mat` (I believe it is boolean flag or smt).

File: modules/gapi/src/backends/sycl/gsyclbackend.cpp [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+
+    const auto bindBuffer = [this](const RcDesc& rc)
+    {   // FIXME: This is likely a broken implementation. Figure out how to perform
+        //        type and dimension matching automatically
+        auto& mag_buffer = m_res.template slot<sycl::buffer<uint8_t, 2>>()[rc.id];
+        auto& mag_mat = m_res.template slot<cv::Mat>()[rc.id];
+        CV_CheckTypeEQ(mag_mat.type(), CV_8UC1, "Only single channel UInt8 Mats " <<
+                                    "are currently supported by SYCL");
+        mag_buffer(mag_mat.data, range<2>(mag_mat.rows, mag_mat.cols));
+    };
```
Comment (dmatveev) [inline modules/gapi/src/backends/sycl/gsyclbackend.cpp]: what does this method do?

I see you obtain `mag_buffer` as a SYCL buffer, but what means `mag_buffer(...)` call? Is it copy? Is it assignment?

File: modules/gapi/src/backends/sycl/gsyclkernel.cpp [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// FIXME: Currently using existing OpenCL interoperability with UMats
+//        instead of pure SYCL
+// FIXME: Add options for controlling device selection
+void cv::GSYCLContext::initSYCLContext()
+{
+    cl::sycl::default_selector def_selector;
+    m_queue = cl::sycl::queue(def_selector);
+    m_context = m_queue.get_context();
```
Comment (dmatveev) [inline modules/gapi/src/backends/sycl/gsyclkernel.cpp]: Shouldn't this be assigned by the backend?

**Review by dmatveev — COMMENTED** [review]
Looks cool! I see all the right changes to implement a backend.

Would like to see how a kernel looks like - and how they work together, for sure!

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (Blair-Johnson) [reply modules/gapi/CMakeLists.txt:37]: Yes, as far as I know, SYCL requires a C++17 compiler that supports it.

File: modules/gapi/cmake/FindSYCL.cmake (hunk 2 above)
Comment (Blair-Johnson) [reply modules/gapi/cmake/FindSYCL.cmake:252]: That's a question for @ngaloppo, he added this to the original interoperability projects that I was working on. I

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp (hunk 5 above)
Comment (Blair-Johnson) [reply modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: I think that we do, because there are sycl::buffers<> in function signatures below and these are sycl specific datatypes. I could be wrong though.

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp (hunk 10 above)
Comment (Blair-Johnson) [reply modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: Yes, the 2 represents the dimension- these template values are currently intended to act as placeholders. I was unsure of how to properly wrangle OpenCV metadata about datatype and dimension into the SYCL template.

File: modules/gapi/include/opencv2/gapi/util/variant.hpp (hunk 11 above)
Comment (Blair-Johnson) [reply modules/gapi/include/opencv2/gapi/util/variant.hpp:31]: Currently, we have been compiling the entire GAPI module with the SYCL compiler (which only supports C++17). It would be great if we could limit this to a smaller subset of the module, but I don't know how that would be accomplished.

File: modules/gapi/src/backends/common/gbackend.hpp (hunk 12 above)
Comment (Blair-Johnson) [reply modules/gapi/src/backends/common/gbackend.hpp]: Your previous comments had suggested this, I just missed it when I was undoing my changes.

File: modules/gapi/src/backends/sycl/gsyclbackend.cpp (hunk 14 above)
Comment (Blair-Johnson) [reply modules/gapi/src/backends/sycl/gsyclbackend.cpp]: Would I be creating RMats with uninitialized buffers here? I didn't think that this constructor had access to runtime data yet. I'm still somewhat unclear on the proper use of RMats.

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (ngaloppo) [reply modules/gapi/CMakeLists.txt:37]: @Blair-Johnson have you tried to turn this to C++11 only?

Comment (Blair-Johnson) [reply modules/gapi/CMakeLists.txt:37]: @ngaloppo yes, the compiler finds many errors in the SYCL source code when compiling with the C++11 standard

File: modules/gapi/cmake/FindSYCL.cmake (hunk 2 above)
Comment (ngaloppo) [reply modules/gapi/cmake/FindSYCL.cmake:252]: Good question. I got this version from @vladimir-paramuzov - I hope he can shed some light on the origin of this file.

File: modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp (hunk 5 above)
Comment (ngaloppo) [reply modules/gapi/include/opencv2/gapi/sycl/gsyclkernel.hpp]: Can they be forward declared? I would think so, since you only have references to SYCL types in this header. 
```cpp
namespace sycl {
    template <typename T, typename U> class buffer;
    class queue;
    class context;
} // namespace sycl
```

File: modules/gapi/cmake/FindSYCL.cmake (hunk 2 above)
Comment (vladimir-paramuzov) [reply modules/gapi/cmake/FindSYCL.cmake:252]: Baseline is taken from here: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5701/diffs + added several modifications.

File: modules/gapi/src/backends/sycl/gsyclbackend.cpp (hunk 16 above)
Comment (Blair-Johnson) [reply modules/gapi/src/backends/sycl/gsyclbackend.cpp]: This has been updated to create a new buffer and assign it to mag_buffer.

**Comment by Blair-Johnson** [comment]
Closing for a fresh PR for RMat integration.
